Catch2 does not provide a helpful message when logical operators are used in an assertion

## Description

Catch Classic used to provide a fairly helpful error message when `&&` or `||` was used in an assertion, as the compilation failed inside a type whose name explained the error. For Catch2, the error message looks like this 

```
test1.cpp(6): error C2676: binary '&&': 'Catch::ExprLhs<int *const &>' does not define this operator or a conversion to a type acceptable to the predefined operator
```

which is not exactly helpful.

### Extra information
This was changed during the rewrite of capture layer, likely as an oversight.

* Catch version: **v2.x.x**

--- a/include/internal/catch_decomposer.h
+++ b/include/internal/catch_decomposer.h
@@ -10,6 +10,7 @@
 
 #include "catch_tostring.h"
 #include "catch_stringref.h"
+#include "catch_meta.hpp"
 
 #include <iosfwd>
 
@@ -141,6 +142,20 @@ namespace Catch {
         template<typename RhsT>
         auto operator <= ( RhsT const& rhs ) -> BinaryExpr<LhsT, RhsT const&> const {
             return { static_cast<bool>(m_lhs <= rhs), m_lhs, "<=", rhs };
+        }
+
+        template<typename RhsT>
+        auto operator && ( RhsT const& ) -> BinaryExpr<LhsT, RhsT const&> const {
+            static_assert(always_false<RhsT>::value,
+            "operator&& is not supported inside assertions, "
+            "wrap the expression inside parentheses, or decompose it");
+        }
+
+        template<typename RhsT>
+        auto operator || ( RhsT const& ) -> BinaryExpr<LhsT, RhsT const&> const {
+            static_assert(always_false<RhsT>::value,
+            "operator|| is not supported inside assertions, "
+            "wrap the expression inside parentheses, or decompose it");
         }
 
         auto makeUnaryExpr() const -> UnaryExpr<LhsT> {

--- a/include/internal/catch_meta.hpp
+++ b/include/internal/catch_meta.hpp
@@ -9,6 +9,8 @@
 #ifndef TWOBLUECUBES_CATCH_META_HPP_INCLUDED
 #define TWOBLUECUBES_CATCH_META_HPP_INCLUDED
 
+#include <type_traits>
+
 template< typename... >
 struct TypeList{};
 
@@ -72,5 +74,8 @@ struct combine
         };
     };
 };
+
+template<typename T>
+struct always_false : std::false_type {};
 
 #endif // TWOBLUECUBES_CATCH_META_HPP_INCLUDED

